File: uploads/src/common/mappers/UploadMapper.ts
import {Upload, UploadType} from "../../mongoose/entities";
import {UploadInfoResponse} from "../types/response";
import {config} from "../../config";

export class UploadMapper {
    public toUploadInfoResponse<Metadata>(upload: Upload<Metadata>): UploadInfoResponse<Metadata> {
        return {
            id: upload.id,
            extension: upload.extension,
            meta: upload.meta,
            mimeType: upload.mimeType,
            originalName: upload.originalName,
            size: upload.size,
            type: upload.type,
            thumbnail: upload.thumbnail && this.toUploadInfoResponse(upload.thumbnail),
            preview: upload.preview && this.toUploadInfoResponse(upload.preview),
            name: upload.name,
            isPreview: upload.isPreview,
            isThumbnail: upload.isThumbnail,
            uri: this.getUploadUri(upload)
        }
    }

    private getUploadUri(upload: Upload<any>): string {
        const apiHost = config.API_HOST;

        switch (upload.type) {
            case UploadType.AUDIO:
                return `${apiHost}/api/v1/uploads/audios/${upload.name}`;
            case UploadType.FILE:
                return `${apiHost}/api/v1/uploads/files/${upload.name}`;
            case UploadType.VIDEO:
                return `${apiHost}/api/v1/uploads/videos/${upload.name}`;
            default:
                return `${apiHost}/api/v1/uploads/images/${upload.name}`;
        }
    }
}

File: user/src/main/kotlin/chatox/user/mapper/UploadMapper.kt
package chatox.user.mapper

import chatox.user.api.response.UploadResponse
import chatox.user.domain.ImageUploadMetadata
import chatox.user.domain.Upload
import chatox.user.domain.User
import chatox.user.messaging.rabbitmq.event.UploadCreated
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component

@Component
class UploadMapper {
    @Value("\${uploads.images}")
    private lateinit var imagesBaseUrl: String

    fun <MetadataType>toUploadResponse(upload: Upload<MetadataType>): UploadResponse<MetadataType> = UploadResponse(
            id = upload.id,
            type = upload.type,
            name = upload.name,
            extension = upload.extenstion,
            mimeType = upload.mimeType,
            meta = upload.meta,
            uri = "${imagesBaseUrl}/${upload.name}",
            originalName = upload.originalName
    )

    fun <MetadataType>fromUploadCreated(uploadCreated: UploadCreated<MetadataType>,
                                        preview: Upload<ImageUploadMetadata>?,
                                        user: User?
    ): Upload<MetadataType> = Upload(
            id = uploadCreated.id,
            previewId = preview?.id,
            meta = uploadCreated.meta,
            extenstion = uploadCreated.extension,
            name = uploadCreated.name,
            isPreview = uploadCreated.isPreview,
            isThumbnail = uploadCreated.isThumbnail,
            mimeType = uploadCreated.mimeType,
            size = uploadCreated.size,
            type = uploadCreated.type,
            userId = user?.id,
            originalName = uploadCreated.originalName
    )
}
